lib/blocks.php

```diff
@@ -296,3 +296,209 @@ function gutenberg_enqueue_editor_block_styles_assets() {
 	}
 	add_action( 'enqueue_block_editor_assets', 'gutenberg_enqueue_editor_block_styles_assets' );
 }
+
+/**
+ * Renders the classNames and styles for blocks
+ *
+ * @param string $block_content Output of the current block.
+ * @param array  $block Block Object.
+ * @return string New block output.
+ */
+function gutenberg_experimental_apply_classnames_and_styles( $block_content, $block ) {
+	// Don't filter template part blocks since we filter the blocks in each template
+	// part individually.
+	if ( 'core/template-part' === $block['blockName'] ) {
+		return $block_content;
+	}
+
+	if ( isset( $block['attrs'] ) ) {
+		// Check what style features the block supports.
+		$block_type = WP_Block_Type_Registry::get_instance()->get_registered( $block['blockName'] );
+		if ( ! $block_type ) {
+			return $block_content;
+		}
+		
+		$supports = gutenberg_experimental_global_styles_get_supported_styles( $block_type->supports );
+
+		// Return early if nothing is supported.
+		if ( count( $supports ) === 0 ) {
+			return $block_content;
+		}
+
+		$colors     = gutenberg_experimental_build_css_colors( $block['attrs'], $supports );
+		$typography = gutenberg_experimental_build_css_typography( $block['attrs'], $supports );
+
+		$extra_classes = array_merge(
+			$colors['css_classes'],
+			$typography['css_classes'],
+			isset( $block['attrs']['className'] ) ? array( $block['attrs']['className'] ) : array(),
+			isset( $block['attrs']['align'] ) ? array( 'has-text-align-' . $block['attrs']['align'] ) : array()
+		);
+		$extra_styles  = (
+			$colors['inline_styles'] ||
+			$typography['inline_styles']
+		) ? esc_attr( $colors['inline_styles'] ) .
+			esc_attr( $typography['inline_styles'] )
+		: '';
+
+		$dom = new DOMDocument( '1.0', 'utf-8' );
+
+		// Suppress warnings from this method from polluting the front-end.
+		// @codingStandardsIgnoreStart
+		if ( ! @$dom->loadHTML( $block_content, LIBXML_HTML_NOIMPLIED | LIBXML_HTML_NODEFDTD | LIBXML_COMPACT ) ) {
+		// @codingStandardsIgnoreEnd
+			return $block_content;
+		}
+
+		$xpath      = new DOMXPath( $dom );
+		$block_root = $xpath->query( '/*' )[0];
+
+		if ( empty( $block_root ) ) {
+			return $block_content;
+		}
+
+		// Merge and dedupe new and existing classes and styles.
+		$new_classes = implode( ' ', array_unique( explode( ' ', ltrim( $block_root->getAttribute( 'class' ) . ' ' ) . implode( ' ', $extra_classes ) ) ) );
+		$new_styles  = implode( ' ', array_unique( explode( ' ', ltrim( $block_root->getAttribute( 'style' ) . ' ' ) . $extra_styles ) ) );
```

Comment:
Do we need the `. ' '` appended inside of the `ltrim()` inside of the second arg of the `explode`?

```suggestion
		$new_classes = implode( ' ', array_unique( explode( ' ', ltrim( $block_root->getAttribute( 'class' ) ) . implode( ' ', $extra_classes ) ) ) );
		$new_styles  = implode( ' ', array_unique( explode( ' ', ltrim( $block_root->getAttribute( 'style' ) ) . $extra_styles ) ) );
```

Reply:
Ah, it's because of the concatenation with `$extra_styles` (after `imploding`) and `$extra_styles`, respectively. How about moving it a level or two up?

```suggestion
		$new_classes = implode( ' ', array_unique( explode( ' ', ltrim( $block_root->getAttribute( 'class' ) ) . ' ' . implode( ' ', $extra_classes ) ) ) );
		$new_styles  = implode( ' ', array_unique( explode( ' ', ltrim( $block_root->getAttribute( 'style' ) ) . ' ' . $extra_styles ) ) );
```

(for `$extra_classes`, see https://github.com/WordPress/gutenberg/pull/23007/files#r441801660)

Reply:
I tested this, but it seems to still be including the extra space at the start when there are no existing styles or classes. The reason for including it in the ltrim() is that it will be trimmed if those values are empty when concatenating (so it won't leave an empty space at the start).